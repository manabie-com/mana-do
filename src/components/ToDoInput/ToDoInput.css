.Todo__creation {
    display: flex;
    height: 40px;
    padding: 10px;
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    box-shadow: 0 1px 0.5px rgb(0 0 0 / 10%);
}

.Todo__input {
    width: 100% !important;
    height: 100% !important;
	outline: none;
    border: none;
    margin-left: 10px;
    margin-right: 10px;
    font-size: 35px;
}

.Todo__input::placeholder {
    opacity: 0.7;
}
