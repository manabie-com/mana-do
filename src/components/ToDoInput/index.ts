export * from './ToDoInput';
export {default} from './ToDoInput';