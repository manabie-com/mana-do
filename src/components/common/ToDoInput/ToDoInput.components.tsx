import React from 'react';
import { useToDoPageContext } from '../../../context/ToDoPageProvider';
import Service from '../../../service';
import { createTodo } from '../../../store/actions';
import Input from '../Input/Input.components';

/* I created a new component to separate the handling logic 
making us have as many stateless components as possible
1. Reusable
2. Capable of maintenance
3. Flexibility
*/
const ToDoInput: React.FC = () => {
  const inputRef = React.useRef<HTMLInputElement>(null);
  const [emptyInput, setEmptyInput] = React.useState<boolean>(false);
  const { dispatch } = useToDoPageContext();
  const handleOnSubmitTodoInput = React.useCallback(
    async (value: string) => {
      // it will good to be optimis
      const resp = await Service.createTodo(value);
      dispatch(createTodo(resp));
    },
    [dispatch]
  );
  const onCreateTodo = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (inputRef.current) {
      const { value } = inputRef.current;
      if (e.key === 'Enter') {
        if (value) {
          handleOnSubmitTodoInput(inputRef.current.value);
          inputRef.current.value = '';
        } else {
          setEmptyInput(true);
        }
      }
      if (value) {
        setEmptyInput(false);
      }
    }
  };
  return (
    <Input
      ref={inputRef}
      className='Todo__input'
      error={emptyInput}
      placeholder={
        emptyInput ? 'This field cannot be left blank' : 'What need to be done?'
      }
      onKeyDown={onCreateTodo}
    />
  );
};

export default React.memo(ToDoInput);
