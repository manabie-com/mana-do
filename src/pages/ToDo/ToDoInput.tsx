import React from 'react';
import { AppContext } from '../../context/toDoContext';
import Service from '../../service';
import { createTodo } from '../../store/actions';

export default function ToDoInput() {
  const inputRef = React.useRef<any>(null);
  const { dispatch } = React.useContext(AppContext);

  React.useEffect(() => {
    inputRef.current.focus();
  }, []);

  const onCreateTodo = async (e: React.KeyboardEvent<HTMLInputElement>) => {
    const { value } = e.currentTarget;
    if (e.key === 'Enter' && value) {
      const resp = await Service.createTodo(inputRef.current.value);
      dispatch(createTodo(resp));

      // It'd be a better user experience if we clear the input every time a to-do is created.
      // So users don't have to delete the old input value before create a new one.
      inputRef.current.value = '';
    }
  };

  return (
    <div className='Todo__creation'>
      <input
        ref={inputRef}
        className='Todo__input'
        placeholder='What need to be done?'
        onKeyDown={onCreateTodo}
      />
    </div>
  );
}
