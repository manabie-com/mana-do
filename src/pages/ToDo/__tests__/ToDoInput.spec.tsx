import React from 'react';
import ToDoInput from '../ToDoInput';
import { render } from '@testing-library/react';

describe('ToDoInput', () => {
  it('Should match snapshot', () => {
    const { container } = render(<ToDoInput />);
    expect(container).toMatchSnapshot();
  });
});

describe('Render ToDoInput', () => {
  it('Should render ToDoInput with placeholder text', async () => {
    const { getByPlaceholderText } = render(<ToDoInput />);
    expect(getByPlaceholderText(/what need to be done/i)).toBeInTheDocument();
  });
});
