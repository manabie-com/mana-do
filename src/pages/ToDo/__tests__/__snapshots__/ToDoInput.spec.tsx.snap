// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoInput Should match snapshot 1`] = `
<div>
  <div
    class="Todo__creation"
  >
    <input
      class="Todo__input"
      placeholder="What need to be done?"
    />
  </div>
</div>
`;
